pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/OffloadIndexBlockImpl.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    public static OffloadIndexBlockImpl get(LedgerMetadata metadata, List<OffloadIndexEntryImpl> entries) {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Sets.newTreeSet();
+        entries.forEach(entry -> block.indexEntries.add(entry));
+        checkState(entries.size() == block.indexEntries.size());
+        block.segmentMetadata = metadata;
+        return block;
+    }
+
+    public static OffloadIndexBlockImpl get(InputStream stream) throws IOException {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Sets.newTreeSet();
+        block.fromStream(stream);
+        return block;
+    }
+
+    public void recycle() {
+        segmentMetadata = null;
+        indexEntries.clear();
+        indexEntries = null;
+        if (recyclerHandle != null) {
+            recyclerHandle.recycle(this);
+        }
+    }
+
+    @Override
+    public OffloadIndexEntry getIndexEntryForEntry(long messageEntryId) {
+        if(messageEntryId > segmentMetadata.getLastEntryId()) {
+            log.warn("Try to get entry: {}, which beyond lastEntryId {}, return null",
+                messageEntryId, segmentMetadata.getLastEntryId());
+            return null;
+        }
+        // find the greatest mapping Id whose entryId <= messageEntryId
+        return this.indexEntries.floor(OffloadIndexEntryImpl.of(messageEntryId, 0, 0));
+    }
+
+    @Override
+    public int getEntryCount() {
+        return this.indexEntries.size();
+    }
+
+    @Override
+    public LedgerMetadata getLedgerMetadata() {
+        return this.segmentMetadata;
+    }
+
+    private static byte[] buildLedgerMetadataFormat(LedgerMetadata metadata) {
+        checkState(metadata instanceof org.apache.bookkeeper.client.LedgerMetadata);
+
+        org.apache.bookkeeper.client.LedgerMetadata ledgerMetadata =
+            (org.apache.bookkeeper.client.LedgerMetadata) metadata;
+
+        return ledgerMetadata.serialize();
+    }
+
+    /**
+     * Get the content of the index block as InputStream.
+     * Read out in format:
+     *   | index_magic_header | index_block_len | index_entry_count |
+     *   |segment_metadata_len | segment metadata | index entries |
+     */
+    @Override
+    public InputStream toStream() throws IOException {
+        int indexBlockLength;
+        int segmentMetadataLength;
+        int indexEntryCount = this.indexEntries.size();
+
+        byte[] ledgerMetadataByte = buildLedgerMetadataFormat(this.segmentMetadata);
+        segmentMetadataLength = ledgerMetadataByte.length;
+
+        indexBlockLength = 4 /* magic header */
+            + 4 /* index block length */
+            + 4 /* segment metadata length */
+            + 4 /* index entry count */
+            + segmentMetadataLength
+            + indexEntryCount * (8 + 4 + 8); /* messageEntryId + blockPartId + blockOffset */
+
+        ByteBuf out = Unpooled.buffer(indexBlockLength, indexBlockLength);
```

Comment:
any reasons why we can't use pooled buffer? I think ByteBufInputStream has a flag to release bytebuf when closing.